RUSTFLAGS override .cargo/config.toml and result in manual workarounds

### `brew doctor` output

```shell
CI run https://github.com/Homebrew/homebrew-core/actions/runs/10208208304/job/28244359400#step:3:264

==> brew doctor
All steps passed!
```


### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
==> brew config
HOMEBREW_VERSION: 4.3.12-52-g53e3632
ORIGIN: https://github.com/Homebrew/brew
HEAD: 53e363258adfced393b6ec99fce66f6ea1ff7687
Last commit: 7 hours ago
Core tap HEAD: e3a6fd41d57b0b82402f33dfb352fd8e2ab23473
Core tap last commit: 3 minutes ago
Core tap JSON: 02 Aug 01:08 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_BOOTSNAP: set
HOMEBREW_CASK_OPTS: []
HOMEBREW_COLOR: set
HOMEBREW_CURL_PATH: /usr/bin/curl
HOMEBREW_DEVELOPER: set
HOMEBREW_FAIL_LOG_LINES: 150
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_GIT_EMAIL: 1589480+BrewTestBot@users.noreply.github.com
HOMEBREW_GIT_NAME: BrewTestBot
HOMEBREW_GIT_PATH: /usr/bin/git
HOMEBREW_LOGS: /__w/homebrew-core/homebrew-core/logs
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_COLOR: set
HOMEBREW_NO_EMOJI: set
HOMEBREW_NO_ENV_HINTS: set
HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: set
HOMEBREW_NO_INSTALL_FROM_API: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.4 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4/bin/ruby
CPU: quad-core 64-bit zen3
Clang: N/A
Git: 2.46.0 => /usr/bin/git
Curl: 7.81.0 => /usr/bin/curl
Kernel: Linux 6.5.0-1024-azure x86_64 GNU/Linux
OS: Ubuntu 22.04.4 LTS
Host glibc: 2.35
/usr/bin/gcc: 11.4.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```


### What were you trying to do (and why)?

Building bottle in CI, e.g. https://github.com/Homebrew/homebrew-core/pull/179360


### What happened (include all command output)?

Failure when building due to overridden [.cargo/config.toml](https://gitlab.com/veilid/veilid/-/blob/v0.3.4/.cargo/config.toml?ref_type=tags#L2):
```
  ==> cargo install --locked --root=/home/linuxbrew/.linuxbrew/Cellar/veilid/0.3.4 --path=veilid-cli
...
  error[E0433]: failed to resolve: could not find `Builder` in `task`
     --> veilid-tools/src/spawn.rs:58:54
      |
  58  |                     MustJoinHandle::new(tokio::task::Builder::new().name(name).spawn(future).unwrap())
      |                                                      ^^^^^^^ could not find `Builder` in `task`
      |
  note: found an item that was configured out
```

### What did you expect to happen?

Successful build without any extra workarounds.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. Remove `ENV["RUSTFLAGS"]` workaround from `veilid`
2. `brew install [--build-from-source | --build-bottle] veilid` on Linux
```

--- a/Library/Homebrew/extend/ENV/shared.rb
+++ b/Library/Homebrew/extend/ENV/shared.rb
@@ -113,6 +113,11 @@ module SharedEnvExtension
     self[key] = PATH.new(self[key]).append(path)
   end
 
+  sig { params(rustflags: String).void }
+  def append_to_rustflags(rustflags)
+    append("HOMEBREW_RUSTFLAGS", rustflags)
+  end
+
   # Prepends a directory to `PATH`.
   # Is the formula struggling to find the pkgconfig file? Point it to it.
   # This is done automatically for keg-only formulae.

--- a/Library/Homebrew/extend/ENV/std.rb
+++ b/Library/Homebrew/extend/ENV/std.rb
@@ -34,7 +34,8 @@ module Stdenv
     self["PKG_CONFIG_LIBDIR"] = determine_pkg_config_libdir
 
     self["MAKEFLAGS"] = "-j#{make_jobs}"
-    self["RUSTFLAGS"] = Hardware.rustflags_target_cpu(effective_arch)
+    self["RUSTC_WRAPPER"] = "#{HOMEBREW_SHIMS_PATH}/super/rustc_wrapper"
+    self["HOMEBREW_RUSTFLAGS"] = Hardware.rustflags_target_cpu(effective_arch)
 
     if HOMEBREW_PREFIX.to_s != "/usr/local"
       # /usr/local is already an -isystem and -L directory so we skip it

--- a/Library/Homebrew/extend/ENV/super.rb
+++ b/Library/Homebrew/extend/ENV/super.rb
@@ -62,7 +62,8 @@ module Superenv
 
     self["HOMEBREW_ENV"] = "super"
     self["MAKEFLAGS"] ||= "-j#{determine_make_jobs}"
-    self["RUSTFLAGS"] = Hardware.rustflags_target_cpu(effective_arch)
+    self["RUSTC_WRAPPER"] = "#{HOMEBREW_SHIMS_PATH}/super/rustc_wrapper"
+    self["HOMEBREW_RUSTFLAGS"] = Hardware.rustflags_target_cpu(effective_arch)
     self["PATH"] = determine_path
     self["PKG_CONFIG_PATH"] = determine_pkg_config_path
     self["PKG_CONFIG_LIBDIR"] = determine_pkg_config_libdir || ""

--- a/Library/Homebrew/shims/super/rustc_wrapper
+++ b/Library/Homebrew/shims/super/rustc_wrapper
@@ -1,0 +1,15 @@
+#!/bin/bash
+
+# Prepend HOMEBREW_RUSTFLAGS to rustc arguments if set.
+# This allows Homebrew to pass optimization flags while still respecting
+# .cargo/config.toml (which sets RUSTFLAGS).
+
+# rustc is passed as the first argument to RUSTC_WRAPPER
+# https://doc.rust-lang.org/cargo/reference/environment-variables.html
+RUSTC="${1}"
+shift
+
+# Prepend HOMEBREW_RUSTFLAGS to the arguments if set.
+# HOMEBREW_RUSTFLAGS is set in extend/ENV/{std,super}.rb
+# shellcheck disable=SC2086,SC2154
+exec "${RUSTC}" ${HOMEBREW_RUSTFLAGS} "$@"

